### PR TITLE
feat(unstable): Deno.ppid

### LINKED ISSUE #6424: Feature Request: Deno.ppid
Similar to `Deno.pid`, but holds the parent process ID.

### PR DESCRIPTION
**Pull request by uki00a**
Closes #6424.
I have no idea on how to get the ppid on Windows, so any help/suggestion will be appreciated!

```typescript
$ cargo run -- doc --builtin Deno.ppid --unstable
    Finished dev [unoptimized + debuginfo] target(s) in 0.16s
     Running `target/debug/deno doc --builtin Deno.ppid --unstable`
Defined in lib.deno.d.ts:4743:2 

const ppid: number
    **UNSTABLE**: New API, yet to be vetted. The pid of the current process's parent.
```

### CODE REVIEW

**Comment by caspervonb** [comment]
So parent_id is available in std::os::unix, might be preferable over nix?

**Comment by MarkTiedemann** [comment]
> I have no idea on how to get the ppid on Windows

AFAIK, you'll have to create a toolhelp snapshot, and iterate over all process entries. Each process entry has both a process ID and a parent process ID, so you can break out of the iteration once you find the Deno process based on the PID.

See https://gist.github.com/mattn/253013/d47b90159cf8ffa4d92448614b748aa1d235ebe4, for a C++ implementation.

See https://github.com/MarkTiedemann/deno_fastlist/blob/master/src/lib.rs for a not-well-tested Rust `winapi` implementation I wrote for a not-yet-stable Deno plugin that needs this capability as well. :)

File: cli/js/lib.deno.ns.d.ts [hunk 1]
```diff
@@ -29,6 +29,10 @@ declare namespace Deno {
   /** The current process id of the runtime. */
   export const pid: number;
 
+  /** The pid of the current process's parent.
+   * Ignored on Windows. */
+  export const ppid: number | null;
```
Comment (ry) [inline cli/js/lib.deno.ns.d.ts:34]: APIs must be added to cli/js/lib.deno.unstable.d.ts first. You can follow the example of 86f92e04c79be81f98e5899638cb6fdb29a4fa64

Comment (ry) [reply cli/js/lib.deno.ns.d.ts:34]: pidParent? parentPid?

Comment (caspervonb) [reply cli/js/lib.deno.ns.d.ts:34]: > pidParent? parentPid?

Given that we already have pid, ppid makes sense no? Kinda odd if one is the POSIX name and the other is not 🤔

File: cli/ops/runtime.rs [hunk 2]
```diff
@@ -78,3 +79,67 @@ fn op_metrics(
...
+  }
+  #[cfg(not(unix))]
+  {
+    ppid_win()
+  }
+}
+
+#[cfg(not(unix))]
+fn ppid_win() -> Value {
+  // Adopted from rustup:
```
Comment (uki00a) [inline cli/ops/runtime.rs:97]: Question: Is this the correct way to adopt the dual licensed code?

File: cli/js/lib.deno.ns.d.ts (hunk 1 above)
Comment (uki00a) [reply cli/js/lib.deno.ns.d.ts]: - I've made `Deno.ppid` unstable.
- I'm keeping the property name for now.

**Comment by uki00a** [comment]
@ry @MarkTiedemann
I added support for windows based on @MarkTiedemann's suggestion and the rustup code: https://github.com/rust-lang/rustup/blob/1.21.1/src/cli/self_update.rs#L1036
Now the CI passes, but I don't have a Windows machine available to test on right now :sob:
I'll ask later on Discord to see if anyone can test it on a Windows machine. (`dev` channel?)

**Comment by MarkTiedemann** [comment]
@uki00a You could add a test similar to this:

```typescript
assertEquals(Deno.pid, parseInt(new TextDecoder().decode(await Deno.run({ cmd: ["deno", "eval", "-p", "--unstable", "Deno.ppid"], stdout: "piped" }).output())));
```

Also I can test it on my Windows machine tomorrow. :)

File: cli/ops/runtime.rs [hunk 3]
```diff
@@ -78,3 +79,67 @@ fn op_metrics(
...
+fn ppid() -> Value {
+  #[cfg(unix)]
+  {
+    use std::os::unix::process::parent_id;
+    serde_json::to_value(parent_id()).unwrap()
+  }
+  #[cfg(not(unix))]
+  {
+    ppid_win()
+  }
```
Comment (MarkTiedemann) [inline cli/ops/runtime.rs:92]: Shouldn't the config be the other way around? e.g. use Unix by default, have Windows be the special case.

Comment (uki00a) [reply cli/ops/runtime.rs:92]: I think using Windows by default is better :thinking:

```rust
fn ppid() -> Value {
  #[cfg(windows)]
  {
    ppid_win()
  }
  #[cfg(not(windows))]
  {
    use std::os::unix::process::parent_id;
    serde_json::to_value(parent_id()).unwrap()
  }
}
```

Comment (MarkTiedemann) [reply cli/ops/runtime.rs:92]: @uki00a Yep, that's exactly what I meant. :)

Comment (uki00a) [reply cli/ops/runtime.rs:92]: Thanks @MarkTiedemann! I've made the changes you pointed out at https://github.com/denoland/deno/pull/6539/commits/2ccdd7b546052450cf9dd1044df0ea51dc553db2.

**Comment by MarkTiedemann** [comment]
@uki00a Just tested on my Windows machine.

```
:: Get pid of current command prompt
C:\dev\deno>powershell (Get-WmiObject Win32_Process -Filter ProcessId=$Pid).ParentProcessId
7228

C:\dev\deno>.\target\debug\deno.exe eval --unstable -p [Deno.pid,Deno.ppid]
[ 2596, 7228 ]
```

Seems to work just fine. :)

**Comment by uki00a** [comment]
Thanks a lot @MarkTiedemann, It realy helps me!

**Comment by MarkTiedemann** [comment]
`Deno.ppid` is still `number | null`.

Are you keeping the `null` because `CreateToolhelp32Snapshot`, `Process32First` or `Process32Next` might fail? From my experience, developing [`fastlist`](https://github.com/MarkTiedemann/fastlist/blob/master/fastlist.cpp), which is included in [ps-list](https://www.npmjs.com/package/ps-list), which is downloaded 400k times per week, these APIs don't fail. There has not been a single bug report ever for any of those failing.

AFAIK, `Process32First` and `Process32Next` fail if you provide the wrong flags to `CreateToolhelp32Snapshot` (e.g. you take a module snapshot instead of a process snapshot) or if you forget to set the `dwSize` of the `PROCESSENTRY32`.

For `CreateToolhelp32Snapshot`, the docs state that _"When taking snapshots that include heaps and modules for a process other than the current process, the `CreateToolhelp32Snapshot` function can fail or return incorrect information for a variety of reasons."_ Note that there is no mention of process snapshots being able to fail.

I think `null` is fine for the race condition case where the parent has exited before the `ppid` has been determined.

It might be nicer to say it's a `number` so people don't have to use `!` whenever they use it, and perhaps throw an `Error` if there was an issue determining the `ppid`?

**Comment by MarkTiedemann** [comment]
> perhaps throw an Error if there was an issue determining the ppid?

Might be best to only throw the error if `Deno.ppid` is actually accessed. Otherwise, if your script doesn't need the ppid anyways, it would be bad if it failed because the parent has exited too quickly and the ppid could not be determined.

Not sure. Maybe `number | null` is fine, too. But in any case, the docs should clearly state why it's possible to be `null`.

**Comment by uki00a** [comment]
@MarkTiedemann 
> Are you keeping the null because CreateToolhelp32Snapshot, Process32First or Process32Next might fail? 

Yes exactly. But I also think this is not ideal.

> Might be best to only throw the error if Deno.ppid is actually accessed. Otherwise, if your script doesn't need the ppid anyways, it would be bad if it failed because the parent has exited too quickly and the ppid could not be determined.

I agree with this :+1: In order to achieve this, I think the following changes are needed:
- Add a new Op that returns ppid. (Let's say `op_ppid`)
- Change `Deno.ppid` from a normal property to a getter.
  - When accessed, it calls `op_ppid` and returns it's result.

```typescript
Object.defineProperty(Deno, {
  ppid: getterOnly(() => sendSync("op_ppid"))
});
```

**Comment by MarkTiedemann** [comment]
Hmm, after thinking about it again, I think I've never seen the snapshot not returning a parent ID for a process entry. I think it may be spoofed (https://blog.f-secure.com/detecting-parent-pid-spoofing/) and it may be incorrect (e.g. parent process has exited, new process has taken its ID), but it should always return a parent process ID.

I have some spare time today so I'll write some C++ code to test what happens if the parent process has exited early.

According to the article linked above, parent process ID spoofing protection requires reading the Windows Event Traces, which I think is out of scope for Deno to detect anyways.

To prevent returning a re-used ID, we could check whether the start time of the parent process is before the start time of the Deno process. But I think this may be out of scope, too.

**Comment by MarkTiedemann** [comment]
Just ran a quick test:

```cpp
// parent.cpp
#include <windows.h>
#include <fstream>
#include <iostream>

int main() {
	DWORD pid = GetCurrentProcessId();
	std::ofstream fs("log.txt");
	fs << "parent.pid:\t" << pid << '\n';
	fs.close();
	STARTUPINFO si;
	si.cb = sizeof(si);
	PROCESS_INFORMATION pi;
	CreateProcess("child.exe", NULL, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi);
	CloseHandle(pi.hProcess);
	CloseHandle(pi.hThread);
	return 0;
}
```

```cpp
// child.cpp
#include <windows.h>
#include <tlhelp32.h>
#include <fstream>
#include <iostream>

int main() {
	Sleep(10000);
	HANDLE snap = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
	PROCESSENTRY32 entry;
	entry.dwSize = sizeof(PROCESSENTRY32);
	Process32First(snap, &entry);
	DWORD pid = GetCurrentProcessId();
	std::ofstream fs;
	fs.open("log.txt", std::ios::app);
	do {
		if (pid == entry.th32ProcessID) {
			fs << "child.pid:\t" << pid << "\nchild.ppid:\t" << entry.th32ParentProcessID << '\n';
			fs.close();
			CloseHandle(snap);
			return 0;
		}
	} while (Process32Next(snap, &entry));
	fs << "child.pid:\t" << pid << "\nchild.ppid:\tN/A" << '\n';
	fs.close();
	CloseHandle(snap);
	return 1;
}
```

```batch
:: test.cmd
@echo off
cl child.cpp 
cl parent.cpp
parent.exe
timeout 11
type log.txt
```

```
C:\>test
:: <removed compiler log>
parent.pid:     6244
child.pid:      13200
child.ppid:     6244
```

So even if the parent is already dead for 10 seconds, the snapshot still returns the correct parent process ID.

So yeah, I think we should document that the parent may be dead and that the parent ID may be reused (and perhaps even that the parent ID may be incorrect because it was spoofed), but I think it's fine to use `number` and assume that there will always be a parent ID.

**Comment by MarkTiedemann** [comment]
Maybe return `-1` if it still fails? That's what Node is doing: https://github.com/nodejs/node/blob/3abb52fdb683c9c9ade1b2c7d16d0f640bbaacfd/deps/uv/src/win/util.c#L361

**Comment by uki00a** [comment]
Like this?
```diff
--- a/cli/js/lib.deno.unstable.d.ts
+++ b/cli/js/lib.deno.unstable.d.ts
@@ -1195,6 +1195,9 @@ declare namespace Deno {
 
   /** **UNSTABLE**: New API, yet to be vetted.
    * The pid of the current process's parent.
+   *
+   * NOTE: Don't assume that the processes identified by this id are always the same.
+   * That process may be dead, or that id may be reused by another process.
    */
-  export const ppid: number | null;
+  export const ppid: number;
 }
```

```diff
--- a/cli/ops/runtime.rs
+++ b/cli/ops/runtime.rs
@@ -113,7 +113,7 @@ fn ppid_win() -> Value {
     // and contains our parent's pid
     let snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
     if snapshot == INVALID_HANDLE_VALUE {
-      return Value::Null;
+      return serde_json::to_value(-1);
     }
 
     let mut entry: PROCESSENTRY32 = mem::zeroed();
@@ -123,7 +123,7 @@ fn ppid_win() -> Value {
     let success = Process32First(snapshot, &mut entry);
     if success == 0 {
       CloseHandle(snapshot);
-      return Value::Null;
+      return serde_json::to_value(-1);
     }
 
     let this_pid = GetCurrentProcessId();
@@ -131,7 +131,7 @@ fn ppid_win() -> Value {
       let success = Process32Next(snapshot, &mut entry);
       if success == 0 {
         CloseHandle(snapshot);
-        return Value::Null;
+        return serde_json::to_value(-1);
       }
     }
     CloseHandle(snapshot);
```

**Comment by MarkTiedemann** [comment]
> NOTE: Don't assume that the process identified by this id is always the same.
> That process may be dead, or that id may be reused by another process.

How about:

> It is not safe to assume that `Deno.ppid` always refers to the pid of the parent process. If the parent process has exited already, its pid may be reused by another process.

**Comment by MarkTiedemann** [comment]
Then again, if I look at the `process.ppid` docs from Node https://nodejs.org/dist/latest-v14.x/docs/api/process.html#process_process_ppid, there is no such warning. So maybe we don't need to bother?

**Comment by uki00a** [comment]
PS:

- Added support for Windows.
  - Windows support is mostly co-authored by @MarkTiedemann.
- `Deno.ppid` is now unstable and non-nullable.
- `Deno.ppid` is set to `-1` when some error occurs.

File: cli/ops/runtime.rs [hunk 4]
```diff
@@ -78,3 +79,67 @@ fn op_metrics(
     "bytesReceived": m.bytes_received
   })))
 }
+
+fn ppid() -> Value {
+  #[cfg(windows)]
+  {
+    ppid_win()
```
Comment (caspervonb) [inline cli/ops/runtime.rs:86]: Nit but maybe just inline this? It's really specific to the implementation.

**Comment by MarkTiedemann** [comment]
@uki00a The current logic loops over all processes, until the process with the current process ID is found. I'm not sure how strong the snapshot guarantees are. If, for example, the current process ID is not in the snapshot, the returned parent process ID would be the one of the last process in the snapshot list. It might be safer to structure the loop as follows:

```
// pseudo code
parent_pid = -1
// ...
loop {
  // ...
  if entry.pid == process_pid {
     parent_pid = entry.parent_pid
     break
  }
}
return parent_pid
```

In this case, `-1` would be returned if the current process ID is not in the snapshot. Then again, I don't know whether it's guaranteed to be in there or not.

**Comment by uki00a** [comment]
@caspervonb @MarkTiedemann @ry @bartlomieju Thanks for your review and help!

**Pull request merged**